README.md
<div align="center">
  <img src="https://raw.githubusercontent.com/meg4cyberc4t/mu_scaling_system/main/.github/icon.png" style="border-radius: 20px" height=100/>
  <br />
  <h1>Mu Scaling System</h1>
  <h3>Make your design adaptive for all screens using Mu system</h3>
</div>

<img src="https://raw.githubusercontent.com/meg4cyberc4t/mu_scaling_system/main/.github/preview.png" style="border-radius: 4%;max-height:500px"/>

### How it works?

When developing applications for different screens, the following problem arises: widgets and text are difficult to adapt to different extensions. For example, if you are developing an application based on the "Iphone 14 Mini", on the "Iphone 14 Pro Max" it will seem small and hard to use. And on tablets it looks disgusting...

This library solves the problem as follows.
Depending on the screen data, you specify a common unit of measurement - Mu. Depending on the screen magnification, you increase the Mu. As a result, almost all widget size actions should be based on Mu.
In total, the specified number of Mu on tablets will be larger, which is why the entire interface will be larger, but since everything depends on Mu, it will be proportional.

### How to prepare it?

#### Option with manual calculation of Mu.
1. Call `MuScaleSystem.instance.updateConditions()` before MaterialApp,  Pass a arguments:
* conditions - list of conditions for Mu.
* defaultValue - value of Mu if all conditions do not fit.

        The conditions work like this:
        Mu is calculated by going through each condition sequentially in the list.
        If Mu finds a condition, it takes a value from it.
        If all the conditions do not fit, Mu is set to defaultValue.
2. Call `MuScaleSystem.instance.updateScaleCalculations()` before MaterialApp, pass a BuildContext.

Example: 
```dart
runApp(
  Builder(
    builder: (BuildContext context) {
      MuScaleSystem.instance.updateConditions(
        <MuCondition>[
          MuCondition.side(
            minShortestSide: 800,
            minLongestSide: 1500,
            mu: 5,
          ),
          MuCondition.side(
            minShortestSide: 1500,
            mu: 6,
          ),
        ],
        defaultValue: 4,
      );
      MuScaleSystem.instance.updateScaleCalculations(context);
      return YourApp();
    },
  ),
);
```

#### Option with auto calculation using MuScalingBuilder:

1. Wrap your MaterialApp in MuScalingBuilder.
2. Pass a arguments in MuScalingBuilder:
* conditions - list of conditions for Mu.
* defaultValue - value of Mu if all conditions do not fit.

        The conditions work like this:
        Mu is calculated by going through each condition sequentially in the list.
        If Mu finds a condition, it takes a value from it.
        If all the conditions do not fit, Mu is set to defaultValue.

Example:
```dart
MuScalingBuilder(
  conditions: [
    MuCondition.side(
        mu: 4,
        minShortestSide: 400,
        maxShortestSide: 800,
    ),
    MuCondition.side(
        mu: 5,
        minShortestSide: 800,
        maxShortestSide: 1500,
    ),
  ],
  defaultValue: 4,
  child: const MyApp(),
),
```

### How to use it?


Use Mu for widgets with num extension:
```dart
SizedBox(
    height: 8.mu,
),
```


Use Mu for TextStyle or TextTheme with extensions:
```dart
TextStyle(fontSize: 14).scale((fontSize, mu) => fontSize * mu / 3.5);
TextTheme().scale((fontSize, mu) => fontSize * mu / 3.5);
```

Important points:
* The function for adapting the size of the text style is not general, you can do your own. It exists in order to adjust the size of the text to Mu, but not to make the text out of size.
* .scale() cannot be used if TextStyle does not have font Size set. (If you want to use out of the box, I recommend the Typography class).

Enjoy!


### Contributing 
Contributions are welcomed!
Here is a curated list of how you can help:
* Report bugs and scenarios that are difficult to implement
* Report parts of the documentation that are unclear
* Fix typos/grammar mistakes
* Update the documentation / add examples
* Implement new features by making a pull-request


CHANGELOG.md
## 0.1.3

* Fix: MuScalingBuilder was rebuilt with each size change

## 0.1.2

* Export MyScaleSystem as an alternative option MuScalingBuilder.

## 0.1.1

* c to the main components
* The Dart SDK version has been downgraded to 2.17.0, Flutter SDK to 3.0.0

## 0.1.0

* Initial version of this package

test/mu_scaling_system_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:mu_scaling_system/mu_scaling_system.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/system.dart
import 'dart:ui';

import 'package:flutter/widgets.dart';
import 'package:mu_scaling_system/src/conditions.dart';

/// Not part of the public api
class MuScaleSystem {
  static final MuScaleSystem instance = MuScaleSystem._();

  MuScaleSystem._();

  int _mu = 1;
  int get mu => _mu;
  int _defaultValue = 1;

  List<MuCondition> _conditions = <MuCondition>[];

  void updateConditions(
    List<MuCondition> conditions, {
    int defaultValue = 1,
  }) {
    _conditions = conditions;
    _mu = _defaultValue = defaultValue;
  }

  void updateScaleCalculations(BuildContext context) {
    final FlutterView currentView = View.of(context);
    final MediaQueryData mediaQueryData = MediaQueryData.fromView(currentView);

    for (final MuCondition condition in _conditions) {
      final int? result = condition.condition(mediaQueryData);
      if (result != null) {
        _mu = result;
        return;
      }
    }
    _mu = _defaultValue;
  }
}


lib/src/builder.dart
import 'package:flutter/widgets.dart';
import 'package:mu_scaling_system/src/conditions.dart';
import 'package:mu_scaling_system/src/system.dart';

/// The main Builder for calculating Mu.
/// Pass [conditions] to it, according to which
/// Mu will be calculated, as well as [defaultValue],
/// if none of the conditions are suitable.
///
/// Important:
/// MuScalingBuilder must be up to MaterialApp,
/// CupertinoApp or WidgetsApp,
/// otherwise the design will not be adaptive to the screen size!
class MuScalingBuilder extends StatefulWidget {
  final Widget child;
  final List<MuCondition> conditions;
  final int defaultValue;

  const MuScalingBuilder({
    required this.child,
    required this.conditions,
    this.defaultValue = 1,
    super.key,
  });

  @override
  State<MuScalingBuilder> createState() => _MuScalingBuilderState();
}

class _MuScalingBuilderState extends State<MuScalingBuilder> {
  @override
  void initState() {
    super.initState();
    MuScaleSystem.instance.updateConditions(
      widget.conditions,
      defaultValue: widget.defaultValue,
    );
  }

  @override
  void didUpdateWidget(covariant MuScalingBuilder oldWidget) {
    super.didUpdateWidget(oldWidget);
    MuScaleSystem.instance.updateConditions(
      widget.conditions,
      defaultValue: widget.defaultValue,
    );
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    MuScaleSystem.instance.updateScaleCalculations(context);
  }

  @override
  Widget build(BuildContext context) => widget.child;
}


lib/src/conditions.dart
import 'package:flutter/widgets.dart';

/// Conditions for calculating Mu.
/// If the condition should not return mu, return null.
/// Used in [MuScalingBuilder].
class MuCondition {
  late final int? Function(MediaQueryData mediaQueryData) condition;

  MuCondition.custom(this.condition);

  MuCondition.fromSize(int? Function(Size size) sizeCondition) {
    condition =
        (MediaQueryData mediaQueryData) => sizeCondition(mediaQueryData.size);
  }

  MuCondition.side({
    required int mu,
    double? minShortestSide,
    double? maxShortestSide,
    double? minLongestSide,
    double? maxLongestSide,
  }) : assert(
          !(minLongestSide == null &&
              maxLongestSide == null &&
              minShortestSide == null &&
              maxShortestSide == null),
          'Conditions are not specified',
        ) {
    condition = (MediaQueryData mediaQueryData) {
      if ((minShortestSide == null ||
              mediaQueryData.size.shortestSide > minShortestSide) &&
          (maxShortestSide == null ||
              mediaQueryData.size.shortestSide < maxShortestSide) &&
          (minLongestSide == null ||
              mediaQueryData.size.longestSide > minLongestSide) &&
          (maxLongestSide == null ||
              mediaQueryData.size.longestSide < maxLongestSide)) {
        return mu;
      }
      return null;
    };
  }
}


lib/src/extensions/text_theme_extension.dart
import 'package:flutter/material.dart';
import 'package:mu_scaling_system/src/extensions/text_style_extension.dart';

extension MuScalingTextThemeExtension on TextTheme {
  TextTheme scale(TextStyleScalingCallback scaleCallback) => copyWith(
        bodyLarge: bodyLarge?.scale(scaleCallback),
        bodyMedium: bodyMedium?.scale(scaleCallback),
        bodySmall: bodySmall?.scale(scaleCallback),
        labelLarge: labelLarge?.scale(scaleCallback),
        displayLarge: displayLarge?.scale(scaleCallback),
        displayMedium: displayMedium?.scale(scaleCallback),
        displaySmall: displaySmall?.scale(scaleCallback),
        headlineMedium: headlineMedium?.scale(scaleCallback),
        headlineSmall: headlineSmall?.scale(scaleCallback),
        titleLarge: titleLarge?.scale(scaleCallback),
        headlineLarge: headlineLarge?.scale(scaleCallback),
        labelMedium: labelMedium?.scale(scaleCallback),
        labelSmall: labelSmall?.scale(scaleCallback),
        titleMedium: titleMedium?.scale(scaleCallback),
        titleSmall: titleSmall?.scale(scaleCallback),
      );
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:mu_scaling_system/mu_scaling_system.dart';

void main() {
  runApp(
    MuScalingBuilder(
      conditions: [
        MuCondition.side(
          mu: 4,
          minShortestSide: 400,
          maxShortestSide: 800,
        ),
        MuCondition.side(
          mu: 5,
          minShortestSide: 800,
          maxShortestSide: 1500,
        ),
      ],
      defaultValue: 4,
      child: const MyApp(),
    ),
  );
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
        textTheme: Typography.material2021()
            .englishLike
            .apply(
              bodyColor: Colors.black,
              displayColor: Colors.black,
              decorationColor: Colors.black,
            )
            .scale((fontSize, mu) => fontSize * mu / 3.5),
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text("mu_scaling_system example"),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'Current shortest side: ${MediaQuery.of(context).size.shortestSide}',
            ),
            SizedBox(height: 2.mu),
            Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(3.mu),
                color: Theme.of(context).colorScheme.inversePrimary,
              ),
              child: Padding(
                padding: EdgeInsets.all(6.mu),
                child: Text(
                  'Mu: ${1.mu}',
                  style: Theme.of(context).textTheme.headlineMedium,
                ),
              ),
            ),
            SizedBox(height: 2.mu),
            Container(
              color: Theme.of(context).colorScheme.inversePrimary,
              width: 210,
              child: Padding(
                padding: EdgeInsets.all(6.mu),
                child: Text(
                  'Real 210 width',
                  style: Theme.of(context).textTheme.headlineMedium,
                ),
              ),
            ),
            SizedBox(height: 2.mu),
            Container(
              color: Theme.of(context).colorScheme.inversePrimary,
              width: 70.mu,
              child: Padding(
                padding: EdgeInsets.all(6.mu),
                child: Text(
                  '70.mu width = ${70.mu}',
                  style: Theme.of(context).textTheme.headlineMedium,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
